Added leader election for mysql-proxy switchboard.

andreas-kupries opened the pull request:
Controlling ticket is https://jira.suse.com/browse/CAP-168

# Notes

- New script doing the readiness check and election
- New package importing kubectl for above
- New properties to configure election: service name, renewal interval.

State stored in annotation of the service.
Winner in writing the annotation is elected.
Renews to hold position. Driven by kube healthchecks.
Expired claim allows for usurpation.

Controlling UAA PR = https://github.com/SUSE/uaa-fissile-release/pull/168
Controlling SCF PR = https://github.com/SUSE/scf/pull/2795

# Testing

- Requires an SCF PR using a bosh release created from this here.

andreas-kupries committed:
[jsc#cap-168] Added leader election for mysql-proxy switchboard.

- New package importing kubectl for below
- New job holding
  - New script doing the readiness check and election
  - New properties to configure election: service name, renewal interval.

State stored in annotation of the service.
Winner in writing the annotation is elected.
Renews to hold position. Driven by kube healthchecks.
Expired claim allows for usurpation.
Logs into a file under /tmp, not an official log, dev only.

andreas-kupries commented:
The command `watch -c 'kubectl describe svc -n cf mysql-proxy-proxy ; echo ; echo 0 ; kubectl describe pod -n cf mysql-proxy-0 | grep skiff-role ; echo 1 ; kubectl describe pod -n cf mysql-proxy-1 | grep skiff-role ; echo'` helps with watching the state of the service, pods, and election. It assumes 2 proxy pods / switchboards.

andreas-kupries committed:
Fix a missing semicolon in fuction definition

mook-as reviewed (commented):
What are the kube guarantees around propagating annotations? That is, if we have a multi-master cluster, can we be sure that the claims will get propagated around correctly? Or do we just run for a bit with multiple active switchboards?

andreas-kupries commented:
I worked from the documentation at https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#annotate with the main sentence of importance to me that __Attempting to set an annotation that already exists will fail unless --overwrite is set__

And the assumption that the backend will use some sort of ACID/consensus (etcd?) to make this happen even for multiple masters.

Googled a bit ... https://matthewpalmer.net/kubernetes-app-developer/articles/how-does-kubernetes-use-etcd.html

andreas-kupries committed:
DRY for name of log file. More logging in the kube actions.

mook-as commented on jobs/switchboard-leader/spec:
```diff
@@ -0,0 +1,19 @@
+description: Job to host the a/p probe for mysql switchboard and leader election
+
+templates:
+  readiness-switchboard.sh: bin/readiness/switchboard
```
This is still an ERB template (it uses `<%`); it should be named `readiness-switchboard.sh.erb` instead.

andreas-kupries replied:
Will fix

andreas-kupries replied:
Done in commit fe21acb49b6b6ea93709501ba6890c59687b9af2

mook-as commented on jobs/switchboard-leader/spec:
```diff
@@ -0,0 +1,19 @@
+
+properties:
+
+  gp.switchboard.service:
```
What is the `gp` prefix?

andreas-kupries replied:
`global-properties`

mook-as replied:
Then please use more explicit naming:
```suggestion
  global-properties.switchboard.service:
```

andreas-kupries replied:
Ok. ... And right, I used defaults for these to avoid having to set anything in the scf and uaa charts. IOW, no excuse tot make the change.

andreas-kupries replied:
Done in commit f3df90d8c255a2be1656283b07fa8257a1bd2080

mook-as commented on jobs/switchboard-leader/spec:
```diff
@@ -0,0 +1,19 @@
+properties:
+
+  gp.switchboard.service:
+    description: Name of the service to work on for a/p leader election.
```
Please avoid acronyms here for clarity: `… for active/passive leader election`

Actually, for even more clarity, we should be explicit: `Name of the Kubernetes service to annotate to determine active/passive leader election.`

andreas-kupries replied:
Will fix

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/spec:
```diff
@@ -0,0 +1,19 @@
+    default: mysql-proxy-proxy
+
+  gp.switchboard.renewal:
+    description: Interval in seconds within which the claim has to be renewed to be kept leader.
```
Should probably say something along the lines of `Should be set to 3 times the readiness probe interval` or something.

andreas-kupries replied:
Yes. Will fix.

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+#!/bin/sh
+# Readiness probe script for the mysql-proxy role
```
This would be a good spot for the algorithm (describing how we use annotations to manage the election).

andreas-kupries replied:
Will do.

andreas-kupries replied:
Done with commit 678674fe397ad162ad75b79787008f89669d09de

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+#!/bin/sh
+# Readiness probe script for the mysql-proxy role
+
+K=/var/vcap/packages/kubectl/bin/kubectl
```
Please do not use single-letter shorthands; they are good for one-off scripts, but get confusing if we have to refer back to things three months later.  For the `kubectl` case in particular, adding it to `$PATH` is probably the better way to go.

andreas-kupries replied:
Will change

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+H="${HOSTNAME}"
+S="<%= p('gp.switchboard.service') %>"
+D="<%= p('gp.switchboard.renewal') %>"
+A=skiff-leader
```
Presumably this stands for `annotation` but this is even more opaque than the other shorthands.

andreas-kupries replied:
Will change

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+S="<%= p('gp.switchboard.service') %>"
+D="<%= p('gp.switchboard.renewal') %>"
+A=skiff-leader
+L=/tmp/log-ready-switchboard
```
Should probably be in `/var/vcap/sys/log`

andreas-kupries replied:
Can do

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+L=/tmp/log-ready-switchboard
+
+# Clear log.
+rm $L
```
`rm -f` please (not needing the `-f` means we forgot to set `errexit`).  While we're at it, set `nounset` too?

andreas-kupries replied:
Will do.

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+
+get_claim () {
+    log retrieval ${N}:${S} :: $A
+    export CLAIM=$($K get service -n $N $S -o "jsonpath={.metadata.annotations.$A}")
```
Please use long option names in scripts: `kubectl get service --namespace $N $S --output=jsonpath=…`

andreas-kupries replied:
Will do

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+}
+
+our_claim () {
+    CLAIMANT=$(echo $CLAIM | awk -F: '{ print $1 }')
```
Also known as `CLAIMANT="${CLAIM%%:*}"`

Also, both here and in other functions: variables should be declared `local` where possible.

andreas-kupries replied:
Ok.

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+}
+
+is_expired () {
+    CLAIMSEC=$(echo $CLAIM | awk -F: '{ print $2 }')
```
Also known as `CLAIMSEC="${CLAIM#*:}"`

andreas-kupries replied:
Ok

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+    CLAIMSEC=$(echo $CLAIM | awk -F: '{ print $2 }')
+    NOW="$(now)"
+    log expired? $NOW :: $CLAIMSEC
+    test -z "${CLAIMSEC}" || test "$(( ${CLAIMSEC} + $D ))" -lt "$NOW"
```
Technically we don't need the first part since `0 + 500` or whatever is definitely going to be less than `$NOW`.

andreas-kupries replied:
`CLAIMSEC` may be the empty string. At that point the expression fails.

mook-as replied:
Actually, in that case:
```bash
$ echo $(( + 500 ))
500
```

andreas-kupries replied:
Ok. Will remove, and hope. I remember having trouble here and I thought it was the empty variable.

andreas-kupries replied:
Done in commit 1202bd7fb7f61e3f8ab6450d714c36403293f1ea

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+claim () {
+    if ! get_claim ; then
+	log no claims
+	make_claim || return 1
```
There's an obvious race here: the claim could have been made after we failed to get it.

We might need to use `kubectl annotate --resource-version` and a bunch of resource version tracking to work around that, but that's definitely more complex code.

andreas-kupries replied:
In that case the `make_claim` will fail, because it does not use `--overwrite`.

mook-as replied:
I think you're right; the races I can think of are covered given that.

mook-as commented on jobs/switchboard-leader/templates/readiness-switchboard.sh:
```diff
@@ -0,0 +1,117 @@
+	# delayed by pretty exactly (D - check interval) seconds.
+	# See (x). So, perform a second check to see if our claim stuck.
+	sleep 1
+	get_claim && our_claim || return 1
```
There's a difference scenario, but also covered by your code:

```
is_expired
              is_expired
clear_claims
make_claim
              clear_claims
              make_claim
```
This is safe because you immediately get the claim again and check that the expected claimant has claimed it.

andreas-kupries replied:
:)

mook-as reviewed (commented):
I just realized one more thing:
`readiness-switchboard.sh` makes way too many logs in normal operation (to a file on disk); we may need to consider if that will fill up the disk too quickly.  Though I guess we may still have logrotate in place so it's fine?

andreas-kupries commented:
>I just realized one more thing: readiness-switchboard.sh makes way too many logs in normal operation (to a file on disk); we may need to consider if that will fill up the disk too quickly. Though I guess we may still have logrotate in place so it's fine?

I am overwriting the log file on each call, see the `rm $L` at the beginning.
The log it makes always shows only the result of the last invokation of the check.

During my tests I used `watch -c cat /path/to/log` to track the action. A `tail -f` may also work, although it will not nicely clear the screen on each round of changes.

andreas-kupries committed:
Addressing comments. Round 1.
- Tweaked descriptions in spec.
- Long option names
- Longer variable names
- Local variable
- Log moved into standard location
- Bash string functions over `echo|awk` combo.
- kubectl into PATH

andreas-kupries committed:
Address comments. Renamed template.

andreas-kupries committed:
Address comments, algorithm explanations.

andreas-kupries commented:
@mook-as Please check. I believe to have addressed everything.

mook-as commented on packages/kubectl/packaging:
```diff
@@ -0,0 +1,16 @@
+
+echo Retrieving kubetcl ${VERSION}
+
+curl -LO https://storage.googleapis.com/kubernetes-release/release/${VERSION}/bin/linux/amd64/kubectl
```
Ideally we'd put the binary in the blobstore, and not download things as part of packaging.

andreas-kupries replied:
I went with what we already had somewhere else. Pretty much a copy.
We can change this in a separate ticket and PR, in all the places.
Fixed the typo I see above, with commit f3df90d8c255a2be1656283b07fa8257a1bd2080

andreas-kupries committed:
Address comment, and fix a separate typo.
